src/charm.py

```diff
@@ -1,19 +1,497 @@
+            event.defer()
+            return
+
+        try:
+            # mongod needs keyFile and TLS certificates on filesystem
+            self._push_keyfile_to_workload(container)
+            self._pull_licenses(container)
+            self._set_data_dir_permissions(container)
+
+        except (PathError, ProtocolError, MissingSecretError) as e:
+            logger.error("Cannot initialize workload: %r", e)
+            event.defer()
+            return
+
+        # Add initial Pebble config layer using the Pebble API
+        container.add_layer(Config.CONTAINER_NAME, self._mongos_layer, combine=True)
+
+        # Restart changed services and start startup-enabled services.
+        container.replan()
+
+    def _on_start(self, event: StartEvent) -> None:
+        """Handle the start event."""
+        # start hooks are fired before relation hooks and `mongos` requires a config-server in
+        # order to start. Wait to receive config-server info from the relation event before
+        # starting `mongos` daemon
+        self.status.set_and_share_status(
+            BlockedStatus("Missing relation to config-server.")
+        )
+
+    def _on_update_status(self, _):
+        """Handle the update status event"""
+        if self.unit.status == Config.Status.UNHEALTHY_UPGRADE:
+            return
+
+        if not self.is_integrated_to_config_server():
+            logger.info(
+                "Missing integration to config-server. mongos cannot run unless connected to config-server."
+            )
+            self.status.set_and_share_status(
+                BlockedStatus("Missing relation to config-server.")
+            )
+            return
+
+        self.status.set_and_share_status(ActiveStatus())
+
+    # END: hook functions
+
+    # BEGIN: helper functions
+    def is_integrated_to_config_server(self) -> bool:
+        """Returns True if the mongos application is integrated to a config-server."""
+        return self.model.relations[Config.Relations.CLUSTER_RELATIONS_NAME] is not None
+
+    def _get_mongos_config_for_user(
+        self, user: MongoDBUser, hosts: Set[str]
+    ) -> MongosConfiguration:
+        return MongosConfiguration(
+            database=user.get_database_name(),
+            username=user.get_username(),
+            password=self.get_secret(APP_SCOPE, user.get_password_key_name()),
+            hosts=hosts,
+            port=Config.MONGOS_PORT,
+            roles=user.get_roles(),
+            tls_external=None,  # Future PR will support TLS
+            tls_internal=None,  # Future PR will support TLS
+        )
+
+    def get_secret(self, scope: str, key: str) -> Optional[str]:
+        """Get secret from the secret storage."""
+        label = generate_secret_label(self, scope)
+        secret = self.secrets.get(label)
+        if not secret:
+            return
+
+        value = secret.get_content().get(key)
+        if value != Config.Secrets.SECRET_DELETED_LABEL:
```

Comment:
Can you explain this? why are we testing against the string `"None"`?

Reply:
I actually cannot, I believe this was initially written by Judit 😅

Reply:
okay - let's please track it in a separate ticket and understand what exactly is happening there for a future PR